### ISSUE TITLE
Review Commonbot Code / Implementation 

### ISSUE THREAD

**Issue by CowMuon** [hicommonwealth/commonwealth#4135]
## Summary 
This is a research spike ticket for reviewing the existing Commonbot code. 

## Background
- Commonbot is Common's Discord bot that was built in fall 2022 (by a contracted resource) mainly with the intent of enabling gating on Discord (replacing an earlier Commonbot prototype built by @timolegros in 2021.) 
- Commonbot can be installed from its home page located at https://bot.commonwealth.im/
- The Github repository for Commonbot is located at https://github.com/hicommonwealth/commonwealth-discord-bot

## Roadmap
As part of Discord Parity initiative we are giving communities the ability to sync content from Discord Forums (which launched January 2023) into Common, unidirectionally to start with, with the anticipation we will be enabling bi-directional sync in the future. 

## Considerations 
Updates will be made in/to the Commonbot code repository which eliminates the need for existing users of the bot (ie they have it running on their Discord) to have to reinstall the bot anew (from the installation page, as above.) TBC, updates to this repository could potentially involve complete teardown/re-write, as long as we're working from that repo (to be confirmed.) 

## Scope 
In preparation for this work we need to review this code which currently no one on the engineering team is familiar with (as it was built by an outside contractor, was not documented, and predated our current code review process. As such we first need to review this code, to better understand it, so that we can plan the change set in a more informed way (and/or as mentioned tear the existing implementation down. 

## Acceptance Criteria
Document germane findings.

**Comment by ianrowan** [comment]
Findings so far are as follows: 

### Architecture 

The discord bot functionality lives across two repositories:
1. https://github.com/hicommonwealth/commonwealth-discord-bot - Hosts bot specific code essentially to send data to/from the bot 
2. https://github.com/hicommonwealth/discord-bot-ui - Hosts API/CW data access layer. This one wasnt mentioned here but this is actually the more crucial piece of the bot as it hosts all business logic for gating and creating actions on the bot itself as well as a UI for managing bot related things

A detailed arch is described in [README.md](https://github.com/hicommonwealth/discord-bot-ui#readme) of the UI repo. 

TL;DR is that a User would click the connect wallet button in the discord bot(repo). This sends a request to the UI(repo) server which checks token balances via TBC and creates callback actions to the server running the bot to assign roles within the guild or channel.

Additionally the discord-bot repo included bot actions to notify of new snapshots to the discord channels. I havent found the service that calls these endpoints yet. 

### Stack 
#### Discord bot

The bot code runs a fastify server, a modern efficient and easy to maintain node.js wrapper, for triggering actions to be sent to the bot in discord via discord.js. The whole server is a single short file: https://github.com/hicommonwealth/commonwealth-discord-bot/blob/main/src/index.ts

Additionally discord UI components are built in https://github.com/hicommonwealth/commonwealth-discord-bot/tree/main/src/events which handle new actions created on discord such as connecting a wallet or creating channels.

#### Discord UI and server 

The stack used for the server is defined in the readme 
<img width="832" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/36428666/4aadbec1-02f9-4ee0-a67c-e73c8af2e11b">

Altogether the UI uses modern web frameworks and concepts and the backend is quite robust for this use case. On this note I thing what is currently there is a great starting point as its all very modern and within the realm of our main CW stack which should make development seamless for our devs. 

Additionally the backend is setup to support Discords Oauth patterns and create user sessions which is a great starting point for any additional discord work. 

### Next Steps

I think it makes alot of sense to build right on top of what is already there for both the bot and the API. We can easily remove features we dont want anymore in the bot repo but given the only 2 features are wallet based role assignment and snapshot notifs I'd guess we can just add on and modify logic on the server side + add more commands to the bot. 

And if I'm understanding the unidirectional sync we want with discord the process is as follows using the current bot. 
1. Bot installed on server
2. Bot added to given channel/forum
3. Our API key now has access to poll the data in any channel the bot is a member of 

So if that is the direction we are going(Discord Forum -> CW) a bot living in any of our communities discord currently would only need to be added to the proper channels and we could handle the polling and data storage on the backend w/o any client side changes

**Comment by jnaviask** [comment]
cc @CowMuon to review and close
